src/sql/workbench/api/node/mainThreadNotebook.ts

```diff
@@ -170,30 +202,259 @@ class ServerManagerWrapper implements sqlops.nb.ServerManager {
+		return this.doChangeKernel(kernelInfo);
+	}
+
+	private async doChangeKernel(kernelInfo: sqlops.nb.IKernelSpec): Promise<sqlops.nb.IKernel> {
+		let kernelDetails = await this._proxy.ext.$changeKernel(this.sessionDetails.sessionId, kernelInfo);
+		this._kernel = new KernelWrapper(this._proxy, kernelDetails);
+		return this._kernel;
+	}
+}
+
+class KernelWrapper implements sqlops.nb.IKernel {
+	private _isReady: boolean = false;
+	private _ready = new Deferred<void>();
+	private _info: sqlops.nb.IInfoReply;
+	constructor(private _proxy: Proxies, private kernelDetails: INotebookKernelDetails) {
+		this.initialize(kernelDetails);
+	}
+
+	private async initialize(kernelDetails: INotebookKernelDetails): Promise<void> {
+		try {
+			this._info = await this._proxy.ext.$getKernelReadyStatus(kernelDetails.kernelId);
+			this._isReady = true;
+			this._ready.resolve();
+		} catch (error) {
+			this._isReady = false;
+			this._ready.reject(error);
+		}
+	}
+
+	get isReady(): boolean {
+		return this._isReady;
+	}
+	get ready(): Thenable<void> {
+		return this._ready.promise;
+	}
+
+	get id(): string {
+		return this.kernelDetails.id;
+	}
+
+	get name(): string {
+		return this.kernelDetails.name;
+	}
+
+	get supportsIntellisense(): boolean {
+		return this.kernelDetails.supportsIntellisense;
+	}
+
+	get info(): sqlops.nb.IInfoReply {
+		return this._info;
+	}
+
+	getSpec(): Thenable<sqlops.nb.IKernelSpec> {
+		return this._proxy.ext.$getKernelSpec(this.kernelDetails.kernelId);
+	}
+
+	requestComplete(content: sqlops.nb.ICompleteRequest): Thenable<sqlops.nb.ICompleteReplyMsg> {
+		return this._proxy.ext.$requestComplete(this.kernelDetails.kernelId, content);
+	}
+
+	requestExecute(content: sqlops.nb.IExecuteRequest, disposeOnDone?: boolean): sqlops.nb.IFuture {
+		let future = new FutureWrapper(this._proxy);
+		this._proxy.ext.$requestExecute(this.kernelDetails.kernelId, content, disposeOnDone)
+		.then(details => {
+			future.setDetails(details), error => future.setError(error);
+			this._proxy.main.addFuture(details.futureId, future);
+		});
+		return future;
+	}
+
+	interrupt(): Thenable<void> {
+		return this._proxy.ext.$interruptKernel(this.kernelDetails.kernelId);
+	}
+}
+
+
+class FutureWrapper implements FutureInternal {
+	private _futureId: number;
+	private _done = new Deferred<sqlops.nb.IShellMessage>();
+	private _messageHandlers = new Map<FutureMessageType, Array<sqlops.nb.MessageHandler<sqlops.nb.IMessage>>>();
+	private _msg: sqlops.nb.IMessage;
+	private _inProgress: boolean;
+
+	constructor(private _proxy: Proxies) {
+        this._inProgress = true;
+	}
+
+	public setDetails(details: INotebookFutureDetails): void {
+		this._futureId = details.futureId;
+		this._msg = details.msg;
+	}
+
+	public setError(error: Error | string): void {
+		this._done.reject(error);
+	}
+
+	public onMessage(type: FutureMessageType, payload: sqlops.nb.IMessage): void {
+		let handlers = this._messageHandlers.get(type);
```

Comment:
I simplified message passing so that just 1 message is sent from exthost -> main thread and then we route to the appropriate handler queue.